build_runner/lib/src/asset_graph/graph.dart

```diff
@@ -344,6 +322,7 @@ class AssetGraph {
           //
           // The regular input hash checks won't pick up glob changes.
           node.state = GeneratedNodeState.definitelyNeedsUpdate;
+          node.isFailure = false;
```

Comment:
Discussed offline. It might make things cleaner if failed is considered another state.

Reply:
I was wrong about needing to reset isFailure. We explicitly need to track this separately from the state, and we can't reset it when the state becomes "needs update", because in the case where the inputs were identical and we decide not to rerun we have to remember that the result is a failure rather than success.

The `definitelyNeedsUpdate` case would have been fine since we'd always rebuild it anyway, but the `mayNeedUpdate` case was getting clobbered by https://github.com/dart-lang/build/blob/c460c4a47baf16f29dedb4382f9260e51d30cf80/build_runner/lib/src/generate/build_impl.dart#L548-L552 so subsequent builds with identical digest were looking like they were successful. I added a test for this case.